api/traces/v1/http.go

```diff
@@ -118,23 +126,74 @@ func NewV2Handler(read *url.URL, readTemplate string, upstreamCA []byte, upstrea
 
 	r := chi.NewRouter()
 
-	var proxyRead http.Handler
-	{
-		level.Debug(c.logger).Log("msg", "Configuring upstream Jaeger", "queryv2", read)
+	middlewareMetrics := proxy.MiddlewareMetrics(c.registry, prometheus.Labels{"proxy": "tracesv1-read"})
+
+	if read != nil || readTemplate != "" {
+
+		var proxyRead http.Handler
+		{
+			level.Debug(c.logger).Log("msg", "Configuring upstream Jaeger", "queryv2", read)
+
+			var upstreamMiddleware proxy.Middleware
+			if read != nil {
+				upstreamMiddleware = proxy.MiddlewareSetUpstream(read)
+			} else {
+				upstreamMiddleware = middlewareSetTemplatedUpstream(c.logger, readTemplate)
+			}
+
+			middlewares := proxy.Middlewares(
+				upstreamMiddleware,
+				proxy.MiddlewareSetPrefixHeader(),
+				proxy.MiddlewareLogger(c.logger),
+				middlewareMetrics,
+			)
+
+			t := &http.Transport{
+				DialContext: (&net.Dialer{
+					Timeout: dialTimeout,
+				}).DialContext,
+				TLSClientConfig: tls.NewClientConfig(upstreamCA, upstreamCert),
+			}
+
+			proxyRead = &httputil.ReverseProxy{
+				Director:  middlewares,
+				ErrorLog:  proxy.Logger(c.logger),
+				Transport: otelhttp.NewTransport(t),
 
-		var upstreamMiddleware proxy.Middleware
-		if read != nil {
-			upstreamMiddleware = proxy.MiddlewareSetUpstream(read)
-		} else {
-			upstreamMiddleware = middlewareSetTemplatedUpstream(c.logger, readTemplate)
+				// This is a key piece, it changes <base href=> tags on text/html content
+				ModifyResponse: jaegerUIResponseModifier,
+			}
 		}
 
-		middlewares := proxy.Middlewares(
-			upstreamMiddleware,
-			proxy.MiddlewareSetPrefixHeader(),
-			proxy.MiddlewareLogger(c.logger),
-			proxy.MiddlewareMetrics(c.registry, prometheus.Labels{"proxy": "tracesv1-read"}),
-		)
+		r.Group(func(r chi.Router) {
+			r.Use(c.readMiddlewares...)
+			r.Get("/api/traces*", c.instrument.NewHandler(
+				prometheus.Labels{"group": "tracesv1api", "handler": "traces"},
+				proxyRead))
+			r.Get("/api/services*", c.instrument.NewHandler(
+				prometheus.Labels{"group": "tracesv1api", "handler": "services"},
+				proxyRead))
+			r.Get("/api/dependencies*", c.instrument.NewHandler(
+				prometheus.Labels{"group": "tracesv1api", "handler": "dependencies"},
+				proxyRead))
+			r.Get("/api/metrics*", c.instrument.NewHandler(
+				prometheus.Labels{"group": "metricsv1api", "handler": "metrics"},
+				proxyRead))
+			r.Get("/static/*", c.instrument.NewHandler(
+				prometheus.Labels{"group": "tracesv1static", "handler": "ui"},
+				proxyRead))
+			r.Get("/search*", c.instrument.NewHandler(
+				prometheus.Labels{"group": "tracesv1ui", "handler": "ui"},
+				proxyRead))
+			r.Get("/favicon.ico", c.instrument.NewHandler(
+				prometheus.Labels{"group": "tracesv1ui", "handler": "ui"},
+				proxyRead))
```

Comment:
Probably we can combine the UI handlers into one via `r.Mount` with `/` as the pattern, similar to how we do here: https://github.com/observatorium/api/blob/main/api/metrics/v1/http.go#L324?

Reply:
One question, if I put the r`.Mount("/"..)` here, Will this take precedence over the other above which is `"/tempo"`? because I wanted that "/tempo" will have a different one, forwarding the request to a different endpoint.

Reply:
Did a quick test and was not able to make it work in that way, even if I put the tempo router first.